recipes/paho-mqtt-cpp/all/conanfile.py

```diff
@@ -1,102 +1,120 @@
-                       "ssl": True
-                       }
-
-    _cmake = None
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "ssl": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+        "ssl": True
+    }
 
     @property
-    def _source_subfolder(self):
-        return "source_subfolder"
+    def _min_cppstd(self):
+        return 11
+
+    def export_sources(self):
+        export_conandata_patches(self)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
     def configure(self):
         if self.options.shared:
-            del self.options.fPIC
+            self.options.rm_safe("fPIC")
 
-        minimal_cpp_standard = "11"
-        if self.settings.compiler.cppstd:
-            tools.check_min_cppstd(self, minimal_cpp_standard)
+        self.options["paho-mqtt-c/*"].shared = self.options.shared
+        self.options["paho-mqtt-c/*"].ssl = self.options.ssl
 
-        self.options["paho-mqtt-c"].shared = self.options.shared
-        self.options["paho-mqtt-c"].ssl = self.options.ssl
+    def layout(self):
+        # src_folder must use the same source folder name the project
+        cmake_layout(self, src_folder="src")
 
     def validate(self):
-        if self.options["paho-mqtt-c"].shared != self.options.shared:
+        if self.settings.compiler.cppstd:
+            check_min_cppstd(self, _min_cppstd)
+        if self.dependencies["paho-mqtt-c"].options.shared != self.options.shared:
             raise ConanInvalidConfiguration("{} requires paho-mqtt-c to have a matching 'shared' option.".format(self.name))
-        if self.options["paho-mqtt-c"].ssl != self.options.ssl:
+        if self.dependencies["paho-mqtt-c"].options.ssl != self.options.ssl:
             raise ConanInvalidConfiguration("{} requires paho-mqtt-c to have a matching 'ssl' option.".format(self.name))
+        if Version(self.version) < "1.2.0" and Version(self.dependencies["paho-mqtt-c"].ref.version) >= "1.3.2":
+            raise ConanInvalidConfiguration("{}/{} requires paho-mqtt-c =< 1.3.1".format(self.name, self.version))
 
     def requirements(self):
-        if tools.Version(self.version) >= "1.2.0":
-            self.requires("paho-mqtt-c/1.3.9")
+        if Version(self.version) >= "1.2.0":
+            self.requires("paho-mqtt-c/1.3.9", transitive_headers=True, transitive_libs=True)
         else:
-            self.requires("paho-mqtt-c/1.3.1") # https://github.com/eclipse/paho.mqtt.cpp/releases/tag/v1.1
+            self.requires("paho-mqtt-c/1.3.1", transitive_headers=True, transitive_libs=True) # https://github.com/eclipse/paho.mqtt.cpp/releases/tag/v1.1
+        # our cmakefiles reference openssl directly with ssl enabled, so we
+        # should directly depend, not just transitively.
+        if self.options.ssl:
+            self.requires("openssl/1.1.1t", direct=False)
```

Comment:
What does it mean ?

Reply:
the dependency on openssl is driven by the dependency in paho-mqtt-c which doesn't transitively expose the openssl libraries to consumers (most don't need it) in conan2 due to changes in dependency management.

Because of this change, paho-mqtt-cpp must explicitly depend on openssl (otherwise it gets no headers) if paho-mqtt-c is using openssl.

In the previous version of the conanfile, the relationship between paho-mqtt-cpp and openssl was implicit.

Reply:
The point of this is not to give users access, there's some confusion about this.

This is strictly for "can this be compiled" so of the headers are not exposed it should not be added

Reply:
The various requires flags are rather unclear (direct in particular), and this is the best I could work out without exceeding my time budgeted to fix this enough to work.

The situation here is rather unusual:  because paho-mqtt-cpp is a wrapper around paho-mqtt-c, the dependency is indirect (it's explicitly dependent on the version that paho-mqtt-c has and isn't exposing), but without a `requires()`, the headers aren't available within it's build environment, and to the best of my knowledge, the final consumers of the wrapper do not need them either, hence why there's no justification to set transitive_headers to True for either package.

Reply:
Poking at this https://github.com/eclipse/paho.mqtt.cpp/blob/2ff3d155dcd10564f1816675789284b4efd79eb7/src/mqtt/ssl_options.h#L30

So paho-mqtt-c should be transitive_headers headers however the transitive libs does not make sense, it's all C code which should be compiled if it provides symbols

As for openssl, direct is an internal implementation detail and there's no use case yet for it being set. So it should removed.

***

My personal experience with these projects, the dependencies on OpenSSL is just a packaging error.

https://github.com/eclipse/paho.mqtt.cpp/blob/2ff3d155dcd10564f1816675789284b4efd79eb7/cmake/FindPahoMqttC.cmake#L29 is just to overcome the publicly exposed headers from 

https://github.com/eclipse/paho.mqtt.c/blob/f7799da95e347bbc930b201b52a1173ebbad45a7/src/SSLSocket.h#L29

Which conan 2.0 handles correctly  but the actual problem is 

https://github.com/conan-io/conan-center-index/blob/6c67eb668424d6a2db0dd39ab4718abe1373a555/recipes/paho-mqtt-c/all/conanfile.py#L59

Should have transitive_headers set to true.

Reply:
Should be fixed with https://github.com/conan-io/conan-center-index/pull/16880

Then we can probably completely axes this option so consumers do not need to worry (in the next pr)

Reply:
```suggestion
        # upstream's cmakefiles reference openssl directly with ssl enabled, so we
        # should directly depend, not just transitively.
        if self.options.ssl:
            self.requires("openssl/1.1.1t")
```